### ISSUE TITLE
unused_assignments lint has trouble tracking that one match guard observes assignment from another match guard.

### ISSUE THREAD

**Issue by pnkfelix** [rust-lang/rust#138069]
I tried this code ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=a48a643b7baed1b398af6074f7262ec4)):

```rust
fn pnk(x: usize) -> &'static str {
    let mut k1 = "k1";
    let mut h1 = "h1";
    match x & 3 {
        3 if { k1 = "unused?"; false } => (),
        _ if { h1 = k1; true } => (),
        _ => (),
    }
    h1
}

fn main() {
    dbg!(pnk(3));
}
```

I expected to see this happen: No lint complaints about the assignment to `k1` in the first match guard being "unused", since the assignment is subsequently observed by the second match guard.

Instead, this happened: The `unused_assignments` lint has a false positive on the code above:

```
warning: value assigned to `k1` is never read
 --> src/main.rs:5:16
  |
5 |         3 if { k1 = "unused?"; false } => (),
  |                ^^
  |
  = help: maybe it is overwritten before being read?
  = note: `#[warn(unused_assignments)]` on by default

warning: `playground` (bin "playground") generated 1 warning
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.89s
     Running `target/debug/playground`
[src/main.rs:13:5] pnk(3) = "unused?"
```

### Meta

I'm testing on the playground, Rust stable 1.85.0

**Comment by pnkfelix** [comment]
Fun fact: This problem seems like it has been present for the entirety of Rust's support for assignments in pattern guards: I see the same false positive from the lint in Rust 1.39. (And in Rust 1.38, the code is rejected because at that time, assignments in pattern guards were illegal).

**Comment by compiler-errors** [comment]
The liveness analysis that powers the unused assignments lint is pretty brittle. Ideally we'd reimplement it on the MIR with a real cfg-aware analysis so that we don't have disagreements between MIR's "uninitialized assignments" and this.
